tests/net/ipv6/src/main.c

```diff
@@ -980,6 +998,49 @@ static void test_change_ll_addr(void)
 		     "Wrong link address 2");
 }
 
+static void test_dad_timeout(void)
+{
+#if defined(CONFIG_NET_IPV6_DAD)
+	struct in6_addr addr1 = { { { 0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+				      0, 0, 0, 0, 0, 0, 0x99, 0x1 } } };
+	struct in6_addr addr2 = { { { 0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+				      0, 0, 0, 0, 0, 0, 0x99, 0x2 } } };
+	struct in6_addr addr3 = { { { 0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+				      0, 0, 0, 0, 0, 0, 0x99, 0x3 } } };
+	struct net_if *iface = net_if_get_default();
+
+	struct net_if_addr *ifaddr;
+
+	expecting_dad = true;
+
+	ifaddr = net_if_ipv6_addr_add(iface, &addr1, NET_ADDR_AUTOCONF, 0xffff);
+	zassert_not_null(ifaddr, "Address 1 cannot be added");
+
+	k_sleep(K_MSEC(10));
+
+	ifaddr = net_if_ipv6_addr_add(iface, &addr2, NET_ADDR_AUTOCONF, 0xffff);
+	zassert_not_null(ifaddr, "Address 2 cannot be added");
+
+	k_sleep(K_MSEC(10));
+
+	ifaddr = net_if_ipv6_addr_add(iface, &addr3, NET_ADDR_AUTOCONF, 0xffff);
+	zassert_not_null(ifaddr, "Address 3 cannot be added");
+
+	k_sleep(K_MSEC(200));
+
+	/* We should have received three DAD queries, make sure they are in
+	 * proper order.
+	 */
+	zassert_true(dad_time[0] < dad_time[1], "DAD timer 1+2 failure");
```

Comment:
I guess for a unit test which runs in a well-known environment (we know that it starts immediately after reboot, thus k_uptime_get_32() will be small, but not zero, and won't wrap-around during test), these checks are good enough ("optimized"). They're not very explicit, e.g. there's not explicit check that all 3 are non-zero (i.e. all 3 were received), but again, inequality tests imply that.